src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -62,34 +62,87 @@
+MTR_DIRECT_MEMBERS
+@interface MTRDeviceDelegateInfo : NSObject {
 @private
-    __weak id _object;
+    void * _delegatePointerValue;
+    __weak id _delegate;
+    dispatch_queue_t _queue;
+    NSArray * _interestedPathsForAttributes;
 }
+
+// Array of interested cluster paths or attribute paths, for attribute report filtering.
+@property (readonly) NSArray * interestedPathsForAttributes;
+
+// Pointer value for logging purpose only
+@property (readonly) void * delegatePointerValue;
+
+- (instancetype)initWithDelegate:(id<MTRDeviceDelegate>)delegate queue:(dispatch_queue_t)queue interestedPathsForAttributes:(NSArray * _Nullable)interestedPathsForAttributes;
+
+// Returns YES if delegate and queue are both non-null, and the block is scheduled to run.
+- (BOOL)callDelegateWithBlock:(void (^)(id<MTRDeviceDelegate>))block;
+
+#ifdef DEBUG
+// Only used for unit test purposes - normal delegate should not expect or handle being called back synchronously.
+- (BOOL)callDelegateSynchronouslyWithBlock:(void (^)(id<MTRDeviceDelegate>))block;
+#endif
+
+// Returns YES if the weak reference no longer points to a live object.
+- (BOOL)delegateIsNull;
+
+// Checks if the otherDelegate argument is the same object that the weak reference points to.
+- (BOOL)containsDelegate:(id<MTRDeviceDelegate>)otherDelegate;
 @end
 
-@implementation MTRWeakReference
-- (instancetype)initWithObject:(id)object
+@implementation MTRDeviceDelegateInfo
+- (instancetype)initWithDelegate:(id<MTRDeviceDelegate>)delegate queue:(dispatch_queue_t)queue interestedPathsForAttributes:(NSArray * _Nullable)interestedPathsForAttributes
 {
     if (self = [super init]) {
-        _object = object;
+        _delegate = delegate;
+        _delegatePointerValue = (__bridge void *) delegate;
+        _queue = queue;
+        _interestedPathsForAttributes = [interestedPathsForAttributes copy];
     }
     return self;
 }
-+ (instancetype)weakReferenceWithObject:(id)object
+
+- (NSString *)description
+{
+    return [NSString stringWithFormat:@"<MTRDeviceDelegateInfo: %p delegate value %p interested paths count %lu>", self, _delegatePointerValue, static_cast<unsigned long>(_interestedPathsForAttributes.count)];
+}
+
+- (BOOL)callDelegateWithBlock:(void (^)(id<MTRDeviceDelegate>))block
 {
-    return [[self alloc] initWithObject:object];
+    id<MTRDeviceDelegate> strongDelegate = _delegate;
+    VerifyOrReturnValue(strongDelegate, NO);
+    dispatch_async(_queue, ^{
+        block(strongDelegate);
+    });
+    return YES;
 }
-- (id)strongObject
+
+#ifdef DEBUG
+- (BOOL)callDelegateSynchronouslyWithBlock:(void (^)(id<MTRDeviceDelegate>))block
 {
-    return _object;
+    id<MTRDeviceDelegate> strongDelegate = _delegate;
+    VerifyOrReturnValue(strongDelegate, NO);
+    
+    block(strongDelegate);
+
+    return YES;
+}
+#endif
+
+- (BOOL)delegateIsNull
+{
+    return (!_delegate || !_queue);
```

Comment:
Will this really do the right thing?  I thought `_delegate` (the weak pointer) will still be non-nil, but trying to assign to a strong pointer will cause it to be nil?  Or is that not how it works?

Reply:
This is treating either "no delegate" or "no queue" as "we have no delegate", however as mentioned above I think we should just hold the queue strongly and only the delegate weakly, and that will simplify this.